cli/compose/template/template_test.go

```diff
@@ -172,3 +172,91 @@ func TestSubstituteWithCustomFunc(t *testing.T) {
 	_, err = SubstituteWith("ok ${NOTHERE}", defaultMapping, pattern, errIsMissing)
 	assert.Check(t, is.ErrorContains(err, "required variable"))
 }
+
+func TestExtractVariables(t *testing.T) {
+	testCases := []struct {
+		dict     map[string]interface{}
+		expected map[string]string
+	}{
+		{
+			dict:     map[string]interface{}{},
+			expected: map[string]string{},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "bar",
+			},
+			expected: map[string]string{},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "$bar",
+			},
+			expected: map[string]string{
+				"bar": "",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar}",
+			},
+			expected: map[string]string{
+				"bar": "",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar?:foo}",
+			},
+			expected: map[string]string{
+				"bar": "",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar?foo}",
+			},
+			expected: map[string]string{
+				"bar": "",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar:-foo}",
```

Comment:
Nested substitutions are not supported currently? (e.g. `${foo:-$bar}`)

Reply:
I'm not actually sure abou that :stuck_out_tongue_closed_eyes: